### ISSUE TITLE
MNIST training accuracy

### ISSUE THREAD

**Issue by nickpezzotti1** [xu-ji/IIC#49]
I am loading the MNIST 685 model. According to `.cluster_eval()` I have around 99% accuracy. Evaluating it manually with the coded listed below returns 85% accuracy. Am I doing something wrong?

1) Load model:
```
config_in = open("/content/code/mnist_original/config.pickle", "rb")
config = pickle.load(config_in)
net = archs.__dict__[config.arch](config)
net.load_state_dict(torch.load("/content/code/mnist_original/best_net.pytorch"))
```
2) Load MNIST:
```
from code.utils.cluster.cluster_eval import cluster_eval
from code.utils.cluster.data import cluster_twohead_create_dataloaders

if "MNIST" in config.dataset:
  sobel = False
else:
  sobel = True

dataloaders_head_A, dataloaders_head_B, \
mapping_assignment_dataloader, mapping_test_dataloader = \
  cluster_twohead_create_dataloaders(config)

net.cuda()
cluster_eval(config, net,
             mapping_assignment_dataloader=mapping_assignment_dataloader,
             mapping_test_dataloader=mapping_test_dataloader,
             sobel=sobel, print_stats=True)
net.cpu()
```

3. Set the mappings from semantic clusters to actual classes as the `.cluster_eval()` recommends:
`mappings = {0: 9, 1: 3, 2: 1, 3: 4, 4: 7, 5: 8, 6: 5, 7: 6, 8: 0, 9: 2}`

4. Manually test for accuracy:
```
batch = next(enumerate(mapping_test_dataloader))[1]
imgs = batch[0]
labels = batch[1]

def calculate_accuracy(imgs, labels, n):
    correct = 0.0
    for i in range(n):
      raw_prediction = torch.argmax(net(imgs[i].reshape(1, 1, 24, 24))[0]).item()
      predicted = mappings.get(raw_prediction, -1)
      label = labels[i].item()

      if int(predicted) == int(label):
          correct += 1
      
    return correct / float(n) 

np.mean(calculate_accuracy(imgs, labels, 700))
```

**Comment by xu-ji** [comment]
Hi, 
`cluster_eval` already includes the computation for accuracy over the whole test set inside. The best sub head accuracy can be found in the `stats_dict` object (https://github.com/xu-ji/IIC/blob/b7470a09a6b8d04eb87bf393cc7e4552acc70692/code/utils/cluster/cluster_eval.py#L355). What is this value?

It seems you are only manually testing the first batch? The accuracy benchmark is computed over the whole test set.

Also, the network should be producing several predictions for each input image when you call `net()`, since the network has several sub-heads. Again, the best sub-head is already found inside the call to `cluster_eval`, I think it's the `best_train_sub_head` field inside `stats_dict`. You should index the result from `net()` by this sub-head. I assume you already got that mapping from `best_train_sub_head_match`.

Also maybe call net.eval() before using the network for testing, just as `cluster_eval` does.

**Comment by nickpezzotti1** [comment]
Hi,

This is the `stats_dict` calculated by `cluster_eval()`:
`{'std': 0.0065171574, 'best_train_sub_head_match': [(0, 9), (1, 3), (2, 1), (3, 4), (4, 7), (5, 8), (6, 5), (7, 6), (8, 0), (9, 2)], 'test_accs': [0.99252856, 0.9791286, 0.9923143, 0.9791, 0.9791286], 'train_accs': [0.99252856, 0.9791286, 0.9923143, 0.9791, 0.9791286], 'best_train_sub_head': 0, 'worst': 0.9791, 'avg': 0.98444, 'best': 0.99252856}`

I also tried testing the other batches and the dataset as a whole and the accuracy doesn't go above 0.86. 

I did indeed get the mapping from `best_train_sub_head_match` in `stats_dict` (`0` in this case).

**Comment by xu-ji** [comment]
So cluster eval gets 0.99252856 which is good. Could you link or include your full script testing the dataset as a whole? It is hard to debug otherwise.

**Comment by nickpezzotti1** [comment]
```
dataloaders_head_A, dataloaders_head_B, \
mapping_assignment_dataloader, mapping_test_dataloader = \
  cluster_twohead_create_dataloaders(config)

batch_itr = enumerate(mapping_test_dataloader)

correct = 0.0
count = 0

for i, batch in batch_itr:
    imgs = batch[0]
    labels = batch[1]  
    for i in range(len(imgs)):
      raw_prediction = torch.argmax(net(imgs[i].reshape(1, 1, 24, 24))[0]).item()
      predicted = mappings.get(raw_prediction, -1)
      label = labels[i].item()

      # Increment counters
      count += 1
      if int(predicted) == int(label):
          correct += 1

print((correct/count), count) # returns 0.84
```

The code can be viewed [here](https://colab.research.google.com/drive/1Lvx1lRqSALz1IAymQQ2MAECT_b_cueaL#scrollTo=BJWnmQbVxobd
).

Viewing the misclassified examples, it seems like it always misclassified 1s. All else is classified reliably but nothing gets assigned to the 1's class. An example can be seen [here](https://colab.research.google.com/drive/1Lvx1lRqSALz1IAymQQ2MAECT_b_cueaL#scrollTo=Y8JPCxGHIrog).

**Comment by xu-ji** [comment]
I think I found the reason. Running images through one at a time messes up the normalisation (batchnorm), because the statistics of the data do not match what the network expects given training. 

Using the original batch size gets 0.992528571429 [(my script)](https://gist.github.com/xu-ji/22f59ce6efb257feb238a33b089a577a). Using smaller batch size of 100 (set `config.batch_sz`) gets 0.992485714286.

**Comment by nickpezzotti1** [comment]
Thanks,

Now I am able to get the desired accuracy. That was the problem.

**Comment by nickpezzotti1** [comment]
Hi, I am writing research on IIC and adversarial attacks and defences. I cannot find many articles describing using batch norm at test time. In particular, if it is used it is used in a dynamic programming way (as this [video](https://www.youtube.com/watch?v=tNIpEZLv_eg) explains), which appears to be different as to how it is done here. 

Am I misinterpreting your comments? Shouldn't the model's accuracy be reported on the accuracy of individual examples?

Thanks,
Nicholas.

**Comment by xu-ji** [comment]
For the MNIST model, config.batchnorm_track is False, which causes track_running_stats to be False in the BatchNorm2d [initialisation](https://github.com/xu-ji/IIC/blob/master/code/archs/cluster/vgg.py#L29). So during testing, the mean and variance is not taken from training time but computed on the fly from the test batch. Hence the test batch needs to be statistically representative of the training batch.

Actually for most of the models config.batchnorm_track is True, so performance doesn't vary with test batch size. I don't recall if there was any particular reason why I didn't set the flag for MNIST, it shouldn't really make a material difference to training.
